File: frontend/src/lib/api/notifications.ts
import { apiClient } from "./client";
import type { Notification, NotificationPreferences } from "./types";

export const notificationsApi = {
  list: (cursor?: string) =>
    apiClient.get<Notification[]>("/api/v1/notifications", {
      params: cursor ? { cursor } : undefined,
    }),

  getUnreadCount: () =>
    apiClient.get<{ count: number }>("/api/v1/notifications/unread-count"),

  markAsRead: (id: number) =>
    apiClient.patch(`/api/v1/notifications/${id}/read`),

  markAllAsRead: () =>
    apiClient.post("/api/v1/notifications/mark-all-read"),

  dismiss: (id: number) =>
    apiClient.delete(`/api/v1/notifications/${id}`),

  getPreferences: () =>
    apiClient.get<NotificationPreferences>("/api/v1/notifications/preferences"),

  updatePreferences: (prefs: Partial<NotificationPreferences>) =>
    apiClient.patch<NotificationPreferences>("/api/v1/notifications/preferences", prefs),
};

File: backend/apps/accounts/urls/notifications.py
"""Notification URL patterns for /api/v1/notifications/."""
from django.urls import path

from apps.accounts import notification_views as views

urlpatterns = [
    # GET /notifications — paginated list
    path("notifications", views.NotificationListView.as_view(), name="notification-list"),

    # GET /notifications/unread-count
    path("notifications/unread-count", views.UnreadCountView.as_view(), name="notification-unread-count"),

    # POST /notifications/mark-all-read
    path("notifications/mark-all-read", views.MarkAllReadView.as_view(), name="notification-mark-all-read"),

    # GET + PATCH /notifications/preferences
    path("notifications/preferences", views.PreferencesView.as_view(), name="notification-preferences"),

    # PATCH /notifications/:id/read
    path("notifications/<int:pk>/read", views.MarkReadView.as_view(), name="notification-mark-read"),

    # DELETE /notifications/:id
    path("notifications/<int:pk>", views.DismissView.as_view(), name="notification-dismiss"),
]
